src/AutoRest.CSharp/Common/Output/Models/Types/SerializableObjectType.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using AutoRest.CSharp.Input;
+using AutoRest.CSharp.Input.Source;
+using AutoRest.CSharp.Output.Builders;
+using AutoRest.CSharp.Output.Models.Serialization.Json;
+using AutoRest.CSharp.Output.Models.Serialization.Xml;
+using AutoRest.CSharp.Output.Models.Types;
+
+namespace AutoRest.CSharp.Common.Output.Models.Types
+{
+    internal abstract class SerializableObjectType : ObjectType
+    {
+        protected SerializableObjectType(BuildContext context) : base(context)
+        {
+        }
+        protected SerializableObjectType(string defaultNamespace, SourceInputModel? sourceInputModel) : base(defaultNamespace, sourceInputModel)
+        {
+        }
+
+        private bool? _includeSerializer;
+        public bool IncludeSerializer => _includeSerializer ??= EnsureIncludeSerializer();
+
+        private bool? _includeDeserializer;
+        public bool IncludeDeserializer => _includeDeserializer ??= EnsureIncludeDeserializer();
+
+        private JsonObjectSerialization? _jsonSerialization;
+        public JsonObjectSerialization? JsonSerialization => HasJsonSerialization ? _jsonSerialization ??= EnsureJsonSerialization() : null;
+
+        private XmlObjectSerialization? _xmlSerialization;
+        public XmlObjectSerialization? XmlSerialization => HasXmlSerialization ? _xmlSerialization ??= EnsureXmlSerialization() : null;
+
+        private bool? _hasJsonSerialization;
+        private bool HasJsonSerialization => _hasJsonSerialization ??= EnsureHasJsonSerialization();
+
+        private bool? _hasXmlSerialization;
+        private bool HasXmlSerialization => _hasXmlSerialization ??= EnsureHasXmlSerialization();
```

Comment:
This is a bit weird.
What is the advantage of this pattern vs the following pattern?
```
protected abstract HasJsonSerialization {get;}
```
This is simpler

Reply:
And if the concern is that we need to ensure that this value will only be calculated once, I believe all the derived class could do this by themselves.

Reply:
This property could be referred multiple times, so even for children we should avoid duplicated calculation.

Reply:
The advantage to the method is it allows for complex logic to be inserted if the answer is not "simple".